apps_download/models/product_product.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2017-Today: Odoo Community Association (OCA)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+
+import os
+import tempfile
+import shutil
+import logging
+import base64
+import subprocess
+import time
+from odoo.exceptions import ValidationError
+from odoo import models, fields, api, _
+_logger = logging.getLogger(__name__)
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    dependent_product_ids = fields.Many2many(
+        'product.product', 'product_product_dependent_rel',
+        'src_id', 'dest_id', string='Dependent Products'
+    )
+    module_path = fields.Char('Module Path')
+
+    @api.constrains('dependent_product_ids')
+    def check_dependent_recursion(self):
+        if not self._check_m2m_recursion('dependent_product_ids'):
+            raise ValidationError(
+                _('Error: You cannot create recursive dependency.')
+            )
+
+    @api.model
+    def child_dependency(self, children):
+        res = self.env['product.product']
+        for child in children:
+            if not child.dependent_product_ids:
+                continue
+            res += child.dependent_product_ids
+            res += self.child_dependency(child.dependent_product_ids)
+        return res
+
+    @api.multi
+    def create_dependency_list(self):
+        ret_val = {}
+        for product in self:
+            ret_val[product.id] = product.dependent_product_ids
+            if product.dependent_product_ids:
+                ret_val[product.id] += self.child_dependency(
+                    product.dependent_product_ids)
+        return ret_val
+
+    @api.multi
+    def generate_zip_file(self):
+        for product in self:
+            if not product.module_path:
+                continue
+            tmp_dir = tempfile.mkdtemp()
+            tmp_dir_2 = tempfile.mkdtemp()
+            dependent_products = product.create_dependency_list()
+            dependent_products = dependent_products[product.id]
+            for dependent_product in dependent_products:
+                if not dependent_product.module_path:
+                    continue
+                subprocess.Popen(['cp', '-r', dependent_product.module_path,
```

Comment:
What about use a python library in order to compatibility with multi-platform? (Maybe shutil)